xray/src/generation.rs

```diff
@@ -676,14 +611,97 @@ pub fn build_xray_quadtree(
+        .try_for_each(|node| -> ImageResult<()> {
+            let strategy: Box<dyn ColoringStrategy> = coloring_strategy_kind.new_strategy();
+            let rect_min = node.bounding_rect.min();
+            let rect_max = node.bounding_rect.max();
+            let min = Point3::new(rect_min.x, rect_min.y, bounding_box.min().z);
+            let max = Point3::new(rect_max.x, rect_max.y, bounding_box.max().z);
+            let bbox = Aabb3::new(min, max);
+            if let Some(image) = xray_from_points(
+                &bbox,
+                Vector2::new(parameters.tile_size_px, parameters.tile_size_px),
+                strategy,
+                parameters,
+            ) {
+                image.save(&get_image_path(&parameters.output_directory, node.id))?;
+                created_leaf_node_ids_tx.send(node.id).unwrap();
+            }
+            progress_bar.lock().unwrap().inc();
+            Ok(())
+        })?;
+    progress_bar.lock().unwrap().finish_println("");
+    drop(created_leaf_node_ids_tx);
+    Ok(created_leaf_node_ids_rx.into_iter().collect())
+}
+
+pub fn create_non_leaf_nodes(
+    created_leaf_node_ids: FnvHashSet<NodeId>,
+    deepest_level: u8,
+    root_level: u8,
+    parameters: &XrayParameters,
+) -> ImageResult<FnvHashSet<NodeId>> {
+    let progress_bar =
+        create_syncable_progress_bar(created_leaf_node_ids.len(), "Assigning background color");
+    let background_color = Rgba::from(parameters.tile_background_color);
+    created_leaf_node_ids
+        .par_iter()
+        .try_for_each(|node_id| -> ImageResult<()> {
+            let image_path = get_image_path(&parameters.output_directory, *node_id);
+            let mut image = image::open(&image_path)?.to_rgba();
+            // Depending on the implementation of the inpainting function above we may get pixels
+            // that are not fully opaque or fully transparent. This is why we choose a threshold
+            // in the middle to consider pixels as background or foreground and could be reevaluated
+            // in the future.
+            image = map_colors(&image, |p| if p[3] < 128 { background_color } else { p });
+            image.save(&image_path)?;
+            progress_bar.lock().unwrap().inc();
+            Ok(())
+        })?;
+    progress_bar.lock().unwrap().finish_println("");
+
+    let mut current_level_nodes = created_leaf_node_ids;
+    let mut all_nodes = current_level_nodes.clone();
+
+    for current_level in (root_level..deepest_level).rev() {
+        current_level_nodes = current_level_nodes
+            .iter()
+            .filter_map(|node| node.parent_id())
+            .collect();
+        build_level(
+            &parameters.output_directory,
+            parameters.tile_size_px,
+            current_level,
+            &current_level_nodes,
+            parameters.tile_background_color,
+        );
+        all_nodes.extend(&current_level_nodes);
+    }
+    progress_bar.lock().unwrap().finish();
```

Comment:
Q: Does `progress_bar.finish()` do anything after we did `progress_bar.finish_println("")` already few lines above?

Reply:
No, I think we can remove this one.